pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go

```diff
@@ -0,0 +1,504 @@
+				shouldReport = true
+			}
+			lggr.Debugw("checked deviation", "key", field.InputKey, "deviationType", field.DeviationType, "currentDeviation", currDeviation.String(), "targetDeviation", field.Deviation.String(), "shouldReport", shouldReport)
+		}
+
+		(*currentState)[field.InputKey] = singleValue
+		if len(field.OutputKey) > 0 {
+			report[field.OutputKey] = singleValue
+		} else {
+			report[field.InputKey] = singleValue
+		}
+	}
+
+	// if SubMapKey is provided, move fields in a nested map
+	if len(a.config.SubMapKey) > 0 {
+		subMap := map[string]any{}
+		for _, field := range a.config.Fields {
+			if field.SubMapField {
+				if len(field.OutputKey) > 0 {
+					subMap[field.OutputKey] = report[field.OutputKey]
+					delete(report, field.OutputKey)
+				} else {
+					subMap[field.InputKey] = report[field.InputKey]
+					delete(report, field.InputKey)
+				}
+			}
+		}
+		report[a.config.SubMapKey] = subMap
+	}
+
+	// if none of the AggregationFields define deviation, always report
+	hasNoDeviation := true
+	for _, field := range a.config.Fields {
+		if field.DeviationType != DEVIATION_TYPE_NONE {
+			hasNoDeviation = false
+			break
+		}
+	}
+	if hasNoDeviation {
+		lggr.Debugw("no deviation defined, reporting")
+		shouldReport = true
+	}
+
+	stateValuesMap, err := values.WrapMap(currentState)
+	if err != nil {
+		return nil, fmt.Errorf("aggregate state wrapmap error: %s", err.Error())
+	}
+	stateBytes, err := proto.Marshal(values.ProtoMap(stateValuesMap))
+	if err != nil {
+		return nil, fmt.Errorf("aggregate state proto marshal error: %s", err.Error())
+	}
+
+	toWrap, err := formatReport(report, a.config.ReportFormat)
+	if err != nil {
+		return nil, fmt.Errorf("aggregate formatReport error: %s", err.Error())
+	}
+	reportValuesMap, err := values.NewMap(map[string]any{
+		a.config.OutputFieldName: toWrap,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("aggregate new map error: %s", err.Error())
+	}
+	reportProtoMap := values.Proto(reportValuesMap).GetMapValue()
+
+	lggr.Debugw("Aggregation complete", "shouldReport", shouldReport)
+
+	return &types.AggregationOutcome{
+		EncodableOutcome: reportProtoMap,
+		Metadata:         stateBytes,
+		ShouldReport:     shouldReport,
+	}, nil
+}
+
+func (a *reduceAggregator) initializeCurrentState(lggr logger.Logger, previousOutcome *types.AggregationOutcome) (*map[string]values.Value, error) {
+	currentState := map[string]values.Value{}
+
+	if previousOutcome != nil {
+		pb := &pb.Map{}
+		proto.Unmarshal(previousOutcome.Metadata, pb)
+		mv, err := values.FromMapValueProto(pb)
+		if err != nil {
+			return nil, fmt.Errorf("initializeCurrentState FromMapValueProto error: %s", err.Error())
+		}
+		err = mv.UnwrapTo(currentState)
+		if err != nil {
+			return nil, fmt.Errorf("initializeCurrentState FromMapValueProto error: %s", err.Error())
+		}
+	}
+
+	zeroValue := values.NewDecimal(decimal.Zero)
+	for _, field := range a.config.Fields {
+		if _, ok := currentState[field.InputKey]; !ok {
```

Comment:
what if InputKey is an empty string?

Reply:
It gets set to
```
{
  "": value
}
```

But that's a good edge case in that if there are two Fields with no input key then they will clash.

Reply:
Since `{"": value}` shouldn't be a valid report, I added a new `reportFormat` for `value` which covers outputting a single value, and I changed it to now hold state as the output structure.